nntrainer/dataset/batch_queue.h

```diff
@@ -91,11 +96,184 @@ class BatchQueue {
+   * @note There is race condition between requesting empty, race condition with
+   * mark_ready should be handled by using MT_safe queue.
+   * @return ScopedView<Sample> sample view. Destroying the returned object will
+   * signal the queue that the sample is filled.
+   */
+  ScopedView<Sample> requestEmpty();
+
+  /**
+   * @brief request filled iteration from the queue.
+   * @note race condition here can be handled by using MT_safe queue
+   * @return ScopedView<Iteration> Ieration view. Destroying the returned object
+   * will signal the queue that the sample is done using.
+   */
+  ScopedView<Iteration> requestFilled();
+
+private:
+  /**
+   * @brief A wrapper object around @c Iteration which marks filled when filling
+   * sample is done
+   * @note the given @a iteration_ and @a bq_ must outleave the lifetime of this
+   * class
+   *
+   */
+  class MarkableIteration {
+  public:
+    /**
+     * @brief Construct a new Markable Iteration object
+     *
+     * @param input_dims input dimensions
+     * @param label_dims label dimensions
+     * @param iq_ iteration queue view to notify
+     */
+    MarkableIteration(const std::vector<ml::train::TensorDim> &input_dims,
+                      const std::vector<ml::train::TensorDim> &label_dims,
+                      IterationQueue *iq);
+
+    /**
+     * @brief Construct a new Markable Iteration object
+     *
+     * @param rhs right side to move
+     */
+    MarkableIteration(MarkableIteration &&rhs);
+
+    /**
+     * @brief Move Assignement operator
+     *
+     * @param rhs rhs to move
+     * @return MarkableIteration& markable iteration
+     */
+    MarkableIteration &operator=(MarkableIteration &&rhs);
+
+    /**
+     * @brief mark iteration that one sample is filled
+     * @todo make this function noexcept
+     */
+    void markSampleFilled() /** noexcept */;
+
+    /**
+     * @brief get underlying iteration
+     *
+     * @return Iteration& iteration
+     */
+    Iteration &get() { return iteration; }
+
+  private:
+    mutable std::mutex notify_mutex;
+    Iteration iteration;
+    IterationQueue *iq;
+    unsigned int num_observed;
+  };
+
+  /**
+   * @brief mark the given iteration filled
+   * @todo make this noexcept with the thread safe queue
+   * @param iteration iteration to mark it as filled
+   */
+  void markFilled(MarkableIteration *iteration) /** noexcept */;
+
+  /**
+   * @brief mark the given iteration empty
+   * @todo make this noexcept with the thread safe queue
+   * @param iteration iteration to mark it as emptied
+   */
+  void markEmpty(MarkableIteration *iteration) /** noexcept */;
+
+  std::vector<MarkableIteration> iterations; /** allocated iterations */
+  MarkableIteration *being_filled; /**< iteration that is being filled now */
+
+  std::vector<Sample>::iterator current_iterator;
+
+  /// @todo use mt safe queue
+  std::queue<MarkableIteration *> empty_q;  /** iterations to be filled */
```

Comment:
both the queues combined will constitute the `vector<MarkableIteration> Iterations`?

Reply:
num_being_filled + empty_q.size() + num_being_used + filled_q.size() will consitute the `vector<MarkableIteration> Iterations`